pkg/signup/signup_service.go

```diff
@@ -0,0 +1,93 @@
+package signup
+
+import (
+	"context"
+	"fmt"
+	crtapi "github.com/codeready-toolchain/api/pkg/apis/toolchain/v1alpha1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+	"strings"
+)
+
+type SignupServiceConfiguration interface {
+	GetNamespace() string
+}
+
+type SignupService interface {
+	CreateUserSignup(ctx context.Context, username, userID string) (*crtapi.UserSignup, error)
+}
+
+type SignupServiceImpl struct {
+	Namespace string
+	Client    UserSignupClient
+}
+
+func NewSignupService(cfg SignupServiceConfiguration) (SignupService, error) {
+	k8sConfig, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := NewUserSignupClient(k8sConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return &SignupServiceImpl{
+		Namespace: cfg.GetNamespace(),
+		Client:    client.UserSignups(cfg.GetNamespace()),
+	}, nil
+}
+
+func (c *SignupServiceImpl) CreateUserSignup(ctx context.Context, username, userID string) (*crtapi.UserSignup, error) {
+	name, err := c.transformAndValidateUserName(username)
+	if err != nil {
+		return nil, err
+	}
+
+	userSignup := &crtapi.UserSignup{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      name,
+			Namespace: c.Namespace,
+		},
+		Spec: crtapi.UserSignupSpec{
+			UserID:        userID,
+			TargetCluster: "",
+			Approved:      false,
+			Username:      username,
+		},
+	}
+
+	created, err := c.Client.Create(userSignup)
+	if err != nil {
+		return nil, err
+	}
+
+	return created, nil
+}
+
+func (c *SignupServiceImpl) transformAndValidateUserName(username string) (string, error) {
+	replaced := strings.ReplaceAll(strings.ReplaceAll(username, "@", "-at-"), ".", "-")
```

Comment:
We have to add something which will prevent creating multiple usersignups for the same user before we integrate this with the Signup controller. Otherwise it will break our signup/login flow right away.
If you prefer to address it in a separate PR then we need to do the following:

- Add two more subtasks to the signup story:
  1) Prevent creating multiple usersignups
  2) Integrate the client with the signup endpoint/controller

`i.` can't be done before `ii.`

Reply:
Two new issues created in JIRA:

- https://jira.coreos.com/browse/CRT-271

- https://jira.coreos.com/browse/CRT-272

Reply:
Both CRT-271 and CRT-272 are about the same problem, aren't they?

So, it's all about:
> Prevent creating multiple usersignups

But what about a JIRA for:
> Integrate the client with the signup endpoint/controller

Reply:
Wouldn't that already be covered by https://jira.coreos.com/browse/CRT-246 ?

Reply:
> Wouldn't that already be covered by https://jira.coreos.com/browse/CRT-246 ?

I don't know :) There is no any details in that JIRA. This PR doesn't include the `POST /signup` controller. This PR introduces the service which is supposed to be used by `POST /signup`. If this integration between `POST /signup` and the UserSignup service from this PR is supposed to be covered in CRT-246 then fine.

Reply:
The business logic for the signup controller's `PostHandler()` function will be handled in part by CRT-245.  There is also the matter of extracting the required claims from the user's access token, which I thought there was supposed to be a subtask for but I can't see it now. I'll follow up on this with Michael.

Reply:
Extracting the claims we need is done. But we do need some glue between the existing reg service code and Shane's stuff. This also includes the async management and user "wait" page.

Reply:
@michaelkleinhenz let's include this glue code as part of CRT-246.